pyop2/fusion.py

```diff
@@ -0,0 +1,1264 @@
+                # 1) Extend function name
+                base_fundecl.name = "%s_%s" % (base_fundecl.name, fuse_fundecl.name)
+                # 2) Concatenate the arguments in the signature
+                base_fundecl.args.extend(fuse_fundecl.args)
+                # 3) Uniquify symbols identifiers
+                fuse_symbols = fuse_info['symbol_refs']
+                for decl in fuse_fundecl.args:
+                    for symbol, _ in fuse_symbols[decl.sym.symbol]:
+                        symbol.symbol = "%s_%d" % (symbol.symbol, unique_id)
+                # 4) Scope and concatenate bodies
+                base_fundecl.children[0] = ast.Block(
+                    [ast.Block(base_fundecl.children[0].children, open_scope=True),
+                     ast.FlatBlock("\n\n// Begin of fused kernel\n\n"),
+                     ast.Block(fuse_fundecl.children[0].children, open_scope=True)])
+            # Eliminate redundancies in the fused kernel's signature
+            self._filter_kernel_args(loops, base_fundecl)
+            # Naming convention
+            fused_ast = base_ast
+            return Kernel(kernels, fused_ast, loop_chain_index)
+
+        fused, fusing = [], [self._loop_chain[0]]
+        for i, loop in enumerate(self._loop_chain[1:]):
+            base_loop = fusing[-1]
+            if base_loop.it_space != loop.it_space or \
+                    (base_loop.is_indirect and loop.is_indirect):
+                # Fusion not legal
+                fused.append((fuse(self, fusing, len(fused)), i+1))
+                fusing = [loop]
+            elif (base_loop.is_direct and loop.is_direct) or \
+                    (base_loop.is_direct and loop.is_indirect) or \
+                    (base_loop.is_indirect and loop.is_direct):
+                # This loop is fusible. Also, can speculative go on searching
+                # for other loops to fuse
+                fusing.append(loop)
+            else:
+                raise RuntimeError("Unexpected loop chain structure while fusing")
+        if fusing:
+            fused.append((fuse(self, fusing, len(fused)), len(self._loop_chain)))
+
+        fused_kernels, offsets = zip(*fused)
+        self._schedule = FusionSchedule(fused_kernels, offsets)
+        self._loop_chain = self._schedule(self._loop_chain)
+
+    def _hard_fuse(self):
+        """Fuse consecutive loops over different iteration sets that do not
+        present RAW, WAR or WAW dependencies. For examples, two loops like: ::
+
+            par_loop(kernel_1, it_space_1,
+                     dat_1_1(INC, ...),
+                     dat_1_2(READ, ...),
+                     ...)
+
+            par_loop(kernel_2, it_space_2,
+                     dat_2_1(INC, ...),
+                     dat_2_2(READ, ...),
+                     ...)
+
+        where ``dat_1_1 == dat_2_1`` and, possibly (but not necessarily),
+        ``it_space_1 != it_space_2``, can be hardly fused. Note, in fact, that
```

Comment:
Are you suggesting the terms "hard" and "soft" don't reflect what's happening? Could you elaborate?